File: app/catalog/caemble_catalog/result_metadata.py
"""Typed, candidate-specific metadata shared by result producers and persistence."""

import math
from collections.abc import Mapping


_INTEGER_BITS = {f"{sign}int{bits}": (sign == "u", bits)
                 for sign in ("", "u") for bits in (8, 16, 32, 64)}
_DTYPES = {"string", "bool", "float16", "float32", "float64", *_INTEGER_BITS}


def validate_metadata_schema(schema, path="metadata"):
    if not isinstance(schema, Mapping):
        raise ValueError(f"{path} must declare named metadata fields")
    for name, field in schema.items():
        where = f"{path}.{name}"
        if not isinstance(name, str) or not name or name in {"__proto__", "constructor", "prototype"}:
            raise ValueError(f"{path} has an invalid field name")
        if not isinstance(field, Mapping) or set(field) - {"dtype", "shape", "quantityKind", "unit", "values"}:
            raise ValueError(f"{where} has an invalid metadata descriptor")
        dtype = field.get("dtype")
        if not isinstance(dtype, str) or dtype not in _DTYPES:
            raise ValueError(f"{where} requires a real numeric, bool or string dtype")
        shape = field.get("shape", ())
        if not isinstance(shape, (list, tuple)) or any(size is not None and (type(size) is not int or size < 0) for size in shape):
            raise ValueError(f"{where}.shape requires nonnegative lengths or null")
        physical = dtype.startswith("float")
        if physical:
            if any(not isinstance(field.get(key), str) or not field[key] for key in ("quantityKind", "unit")):
                raise ValueError(f"{where} requires quantityKind and unit")
        elif "quantityKind" in field or "unit" in field:
            raise ValueError(f"{where} non-floating metadata cannot declare physical units")
        if "values" in field and (dtype != "string" or not isinstance(field["values"], (list, tuple))
                                  or not field["values"] or any(not isinstance(value, str) for value in field["values"])
                                  or len(set(field["values"])) != len(field["values"])):
            raise ValueError(f"{where}.values requires distinct string choices")


def validate_result_metadata(schema, values, path="metadata", *, allow_extra=False):
    validate_metadata_schema(schema, path)
    if not isinstance(values, Mapping) or set(schema) - set(values) or (not allow_extra and set(values) - set(schema)):
        raise ValueError(f"{path} fields differ from their declared metadata schema")
    for name, field in schema.items():
        raw = values[name]
        if hasattr(raw, "tolist"):
            raw = raw.tolist()
        shape, dtype = field.get("shape", ()), field["dtype"]
        dimensions = {}

        def validate(item, depth, where):
            if depth < len(shape):
                if not isinstance(item, (list, tuple)) or (shape[depth] is not None and len(item) != shape[depth]):
                    raise ValueError(f"{where} shape differs from its metadata declaration")
                if depth in dimensions and dimensions[depth] != len(item):
                    raise ValueError(f"{where} metadata arrays must be rectangular")
                dimensions[depth] = len(item)
                for index, child in enumerate(item):
                    validate(child, depth + 1, f"{where}[{index}]")
                return
            if dtype == "string":
                valid = isinstance(item, str) and ("values" not in field or item in field["values"])
            elif dtype == "bool":
                valid = type(item) is bool
            elif dtype in _INTEGER_BITS:
                unsigned, bits = _INTEGER_BITS[dtype]
                lower, upper = (0, 2**bits - 1) if unsigned else (-2**(bits - 1), 2**(bits - 1) - 1)
                valid = type(item) is int and lower <= item <= upper and abs(item) <= 2**53 - 1
            else:
                limit = {"float16": 65504., "float32": 3.4028234663852886e38, "float64": float("inf")}[dtype]
                valid = type(item) in (int, float) and math.isfinite(item) and abs(item) <= limit
            if not valid:
                raise ValueError(f"{where} must match its {dtype} metadata declaration")

        validate(raw, 0, f"{path}.{name}")

File: app/ui/src/contracts/resultMetadata.ts
/** Result meaning is declared once; candidate-specific values travel with each tensor. */
export type ResultMetadataField = Readonly<{
  dtype:
    | 'string'
    | 'bool'
    | 'float16'
    | 'float32'
    | 'float64'
    | 'int8'
    | 'int16'
    | 'int32'
    | 'int64'
    | 'uint8'
    | 'uint16'
    | 'uint32'
    | 'uint64'
  shape?: readonly (number | null)[]
  quantityKind?: string
  unit?: string
  values?: readonly string[]
}>
export type ResultMetadataSchema = Readonly<Record<string, ResultMetadataField>>
export type ResultMetadataValue = string | boolean | number | readonly ResultMetadataValue[]
export type ResultMetadata = Readonly<Record<string, ResultMetadataValue>>

export function assertMetadataSchema(value: unknown): asserts value is ResultMetadataSchema {
  if (!value || typeof value !== 'object' || Array.isArray(value)) throw new Error('Invalid result metadata schema.')
  for (const [name, raw] of Object.entries(value)) {
    if (
      !name ||
      ['__proto__', 'constructor', 'prototype'].includes(name) ||
      !raw ||
      typeof raw !== 'object' ||
      Array.isArray(raw)
    )
      throw new Error('Invalid result metadata field.')
    const field = raw as ResultMetadataField
    if (
      Object.keys(field).some((key) => !['dtype', 'shape', 'quantityKind', 'unit', 'values'].includes(key)) ||
      typeof field.dtype !== 'string' ||
      !/^(string|bool|float(16|32|64)|u?int(8|16|32|64))$/u.test(field.dtype) ||
      (field.shape !== undefined &&
        (!Array.isArray(field.shape) ||
          field.shape.some((size) => size !== null && (!Number.isSafeInteger(size) || size < 0))))
    )
      throw new Error(`Invalid metadata descriptor ${name}.`)
    if (
      field.dtype.startsWith('float')
        ? typeof field.quantityKind !== 'string' || !field.quantityKind || typeof field.unit !== 'string' || !field.unit
        : field.quantityKind !== undefined || field.unit !== undefined
    )
      throw new Error(`Metadata ${name} has inconsistent physical units.`)
    if (
      field.values !== undefined &&
      (field.dtype !== 'string' ||
        !Array.isArray(field.values) ||
        !field.values.length ||
        field.values.some((item) => typeof item !== 'string') ||
        new Set(field.values).size !== field.values.length)
    )
      throw new Error(`Metadata ${name} requires distinct string choices.`)
  }
}

export function assertResultMetadata(
  schema: ResultMetadataSchema | undefined,
  value: unknown,
  path = 'Result metadata',
): asserts value is ResultMetadata | undefined {
  if (schema === undefined && value === undefined) return
  assertMetadataSchema(schema)
  if (!schema || !value || typeof value !== 'object' || Array.isArray(value))
    throw new Error(`${path} requires its declared metadata fields.`)
  const actual = value as Record<string, unknown>
  if (
    Object.keys(actual).length !== Object.keys(schema).length ||
    Object.keys(schema).some((name) => !Object.prototype.hasOwnProperty.call(actual, name))
  )
    throw new Error(`${path} fields differ from their declared metadata schema.`)
  for (const [name, field] of Object.entries(schema)) {
    if (['__proto__', 'constructor', 'prototype'].includes(name)) throw new Error(`${path} has an invalid field name.`)
    const shape = field.shape ?? []
    const dimensions = new Map<number, number>()
    const validate = (item: unknown, depth: number): boolean => {
      if (depth < shape.length) {
        if (
          !Array.isArray(item) ||
          (shape[depth] !== null && item.length !== shape[depth]) ||
          (dimensions.has(depth) && dimensions.get(depth) !== item.length)
        )
          return false
        dimensions.set(depth, item.length)
        return item.every((child) => validate(child, depth + 1))
      }
      if (field.dtype === 'string') return typeof item === 'string' && (!field.values || field.values.includes(item))
      if (field.dtype === 'bool') return typeof item === 'boolean'
      if (typeof item !== 'number' || !Number.isFinite(item)) return false
      if (field.dtype.startsWith('float')) {
        const limit = field.dtype === 'float16' ? 65504 : field.dtype === 'float32' ? 3.4028234663852886e38 : Infinity
        return Math.abs(item) <= limit
      }
      const integer = /^(u?)int(8|16|32|64)$/u.exec(field.dtype)
      if (!integer || !Number.isSafeInteger(item)) return false
      const bits = Number(integer[2]),
        unsigned = integer[1] === 'u'
      return item >= (unsigned ? 0 : -(2 ** (bits - 1))) && item < 2 ** (bits - Number(!unsigned))
    }
    if (!validate(actual[name], 0)) throw new Error(`${path}.${name} differs from its metadata declaration.`)
  }
}

export function cloneResultMetadata(value: ResultMetadata): ResultMetadata {
  const clone = (item: ResultMetadataValue): ResultMetadataValue =>
    Array.isArray(item) ? Object.freeze(item.map(clone)) : item
  return Object.freeze(Object.fromEntries(Object.entries(value).map(([name, item]) => [name, clone(item)])))
}
